### PR TITLE
Add comment display limit and comment deletion

### PR DESCRIPTION
**Pull request by avagarcia0**
Adds an option for choosing the maximum number of comments to display on the page, and adds a button that deletes all currently stored comments. This is step 6 for week 3.

I'll work on formatting the page properly next. In the meantime, here's what it currently looks like:
image

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -39,7 +39,20 @@ <h1>Ava Garcia's Portfolio</h1>
         <button type="submit">Submit</button>
       </form>
 
+      <label for="num-comments">Number of comments to display:</label>
+      <select name="num-comments" id="num-comments" onchange="displayComments()">
+        <option value="-1">All</option>
```
Comment (bryyeh) [inline portfolio/src/main/webapp/index.html]: Can we set the value to be "All"? It seems hacky use "-1" to represent "All".

Also, it looks like getNumCommentsToDisplay() would consider an input of "-1" to be an error.

File: portfolio/src/main/java/com/google/sps/servlets/DeletionServlet.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      datastore.delete(entity.getKey());
+    }
```
Comment (bryyeh) [inline portfolio/src/main/java/com/google/sps/servlets/DeletionServlet.java]: There is an API to delete an Iterable of keys so you don't need to for loop.

https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#delete_2

**Review by bryyeh — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeletionServlet.java (hunk 2 above)
Comment (bryyeh) [reply portfolio/src/main/java/com/google/sps/servlets/DeletionServlet.java]: I realized you'd still have to loop to call getKey() so feel free to ignore this. In general, delete operations are expensive so we like to delete everything in one batch call.

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (avagarcia0) [reply portfolio/src/main/webapp/index.html]: Is the change I made good?

File: portfolio/src/main/java/com/google/sps/servlets/DeletionServlet.java (hunk 2 above)
Comment (acheng1) [reply portfolio/src/main/java/com/google/sps/servlets/DeletionServlet.java]: Making a single call to datastoreService will be more efficient. You can use Iterables.transform [1] to transform each entity to its key.

`Iterable<Key> keysToDelete = Iterables.transform(results.asIterable(), Entity::getKey);
datastore.delete(keysToDelete);`

[1] https://guava.dev/releases/21.0/api/docs/com/google/common/collect/Iterables.html#transform-java.lang.Iterable-com.google.common.base.Function-

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -49,6 +49,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       comments.add(body);
     }
 
+    int numComments = getNumCommentsToDisplay(request);
+
+    if (numComments >= 0 && numComments <= comments.size()) {
+      comments = comments.subList(0, numComments);
+    }
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The `PreparedQuery` has an `asList` method [1] that takes `FetchOptions` [2] object. You can use the the fetch options to specify a limit so that you don't get more results back than you need.

[1] https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery.html#asList-com.google.appengine.api.datastore.FetchOptions-
[2] https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/FetchOptions.html

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -82,4 +88,33 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
   }
+
+  /**
+   * @return the number of comments to display, or -1 if all comments should be displayed
+   */
+  private int getNumCommentsToDisplay(HttpServletRequest request) {
+    // Get the input from the form.
+    String numCommentsString = getParameter(request, "num-comments", null);
+
+    if (numCommentsString.equals("All")) {
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:100]: This if fine for this project, but generally I would consider always having some limit :) If this gets to be too many comments, I might consider pagination.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -82,4 +88,33 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
+      return -1;
+    }
+
+    // Convert the input to an int.
+    int numComments;
+    try {
+      numComments = Integer.parseInt(numCommentsString);
+    } catch (NumberFormatException e) {
+      System.err.println("Could not convert to int: " + numCommentsString);
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:109]: This is fine as well, but you may consider using a logger over System.err/out which offers more flexibility like log levels. There are a few options like java.util.logging or Log4j.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -82,4 +88,33 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
+      return -1;
+    }
+
+    // Check that the input is non-negative.
+    if (numComments < 0) {
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:114]: Good test for edge cases!

Comment (avagarcia0) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:114]: Thanks!

**Review by bryyeh — APPROVED** [review]

**Review by acheng1 — APPROVED** [review]
